### ISSUE TITLE
GraphQL Language Features

### ISSUE THREAD

**Issue by jahewson** [jahewson/elm-graphql#2]
Still to do:

- [ ] Fragments (these could become type aliases?)
- [ ] ID types (i.e. references)
- [ ] Mutation
- [ ] Directives
- [ ] Default values (for parameters - use a Maybe?)
- [ ] Aliases

**Comment by carleryd** [comment]
Is anyone working on any of these features? I'm looking into implementing Mutations because the project I'm working on desperately needs this feature. If anyone has any pointers, that would be greatly appreciated, I'm on the elmlang-slack :)

**Comment by jahewson** [comment]
I've been a bit backlogged recently. I should get some time this weekend. The wiki contains a few links and explanations if you're wanting to learn more.

**Comment by jahewson** [comment]
Fragments implemented via 561c843021db626d442269bc220391805f586f96

**Comment by jahewson** [comment]
ID type implemented via fdfd70b045efc7f3e9abd2d4a1e74cf0f6d8ba71

**Comment by jahewson** [comment]
Mutations implemented via 363c6eb7efe611b4bb3653ad6aa82ac3521001e2

**Comment by jahewson** [comment]
Default values implemented via 31cc123178f5ad15c00dac90b4640932873fe60f

**Comment by jahewson** [comment]
Non-leaf input values implemented via e63d966e025036fb7397e574ff07fe74db402c82.

**Comment by note89** [comment]
YEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE :D

**Comment by jahewson** [comment]
Unions implemented via 20709c3717a4c3ed5d2b6511cf2416013468f1c9.

**Comment by liamcurry** [comment]
Can we add [enums](https://facebook.github.io/graphql/#sec-Enums) to this list?

**Comment by jahewson** [comment]
Sure thing!

**Comment by liamcurry** [comment]
@jahewson Have they? I'm getting this error when I try to use an an enum as input:

```
/home/sagan/Dropbox/Projects/dbp/node_modules/elm-graphql/lib/query-to-elm.js:294
            throw new Error('not implemented: ' + type.constructor.name);
                  ^
Error: not implemented: GraphQLEnumType
    at encoderForInputType (/home/sagan/Dropbox/Projects/dbp/node_modules/elm-graphql/lib/query-to-elm.js:294:19)
```

Seems like it needs to be added to [encoderForInputType](https://github.com/jahewson/elm-graphql/blob/master/src/query-to-elm.ts#L329)?

**Comment by liamcurry** [comment]
@jahewson I created PR #35 to try and fix this. I'm happy to make changes if you see anything wrong so just let me know.

**Comment by jahewson** [comment]
Ah, I must have broken them during refactoring. Need more tests!
